### PR TITLE
MWPW-129129: Improve user entitlements

### PR DESCRIPTION
**Pull request by mokimo**
### Description
We got some feedback for https://github.com/adobecom/milo/pull/522 the default response for the getUserEntitlements utility now looks similar to what we provide in the dexter world with `feds.utilities.getUserEntitlements().then(data => console.log(data))` 

An empty response will be provided in any case of error instead of rejecting the promise.
```js
// Example output import getEntitlements from './utilities/getUserEntitlements.js';
const res = await getEntitlements({params, locale});
res = {
    "clouds": {
        "creative_cloud": false,
        "document_cloud": false,
    },
    "arrangment_codes": {
        "arrangment_codes-1": true,
        "arrangment_codes-2": true
    },
    "fulfilled_codes": {
        "fulfilled_codes-1": true,
        "fulfilled_codes-2": true,
    },
    "offer_families": {
        "offer_families_string": true
    },
    "list": {
        "fulfilled_codes": [
            "fulfilled-code-string-1",
            "fulfilled-code-string-2",
        ]
    }
}
```

A raw response can still be retrieved
```js
// Example output import getEntitlements from './utilities/getUserEntitlements.js';
// const res = await getEntitlements({params, locale, format: 'raw'});
// res = raw/unformatted output from the JIL api from https://www.adobe.com/aos-api/users/${profile.userId}/subscriptions
```
Resolves: [MWPW-129129](https://jira.corp.adobe.com/browse/MWPW-129129)

**Test URLs:**
Not required, no live changes

### CODE REVIEW

**Comment by mokimo** [comment]
@yesil would be lovely if you can have a look if that's in line with what we discussed / what you need.

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 1]
```diff
@@ -63,27 +130,37 @@ const getSubscriptions = async ({ queryParams, locale }) => {
...
-      reject = _reject;
-    });
-  setTimeout(() => reject('Subscriptions API Call timeout'), API_WAIT_TIMEOUT);
+   || new Promise((_resolve) => { resolve = _resolve; });
+
+  setTimeout(() => resolve(emptyEntitlements()), API_WAIT_TIMEOUT);
   // TODO we might need to format data for analytics
-  const data = await getSubscriptions({ queryParams, locale });
+  const data = await getSubscriptions({ queryParams, locale })
+    .finally((res) => res || emptyEntitlements());
```
Comment (yesil) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:154]: as per, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally

finally does not take any argument, `res` is not useful here.

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 2]
```diff
@@ -63,27 +130,37 @@ const getSubscriptions = async ({ queryParams, locale }) => {
...
   const queryParams = getQueryParameters(params);
-  if (entitlements[queryParams]) return entitlements[queryParams];
+  if (entitlements[queryParams]) {
+    return format === 'raw'
+      ? entitlements[queryParams]
+      : entitlements[queryParams]
+        .then((res) => mapSubscriptionCodes(res));
+  }
 
   let resolve;
```
Comment (yesil) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:147]: looks like concurrent calls might override resolve object.
Why not move `setTimeout` logic inside the promise and remove `resolve` variable?

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js (hunk 1 above)
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js]: Indeed, changed that 👍

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js (hunk 2 above)
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js]: Moved a bit into the promise itself and added an abort controller for the setTimeout, feel like that's a little cleaner

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 3]
```diff
@@ -54,7 +124,10 @@ const getSubscriptions = async ({ queryParams, locale }) => {
         'Accept-Language': getAcceptLanguage(locale).join(','),
       },
     })
-    .then((response) => (response.status === 200 ? response.json() : null));
+    .then((response) => {
+      clearTimeout(timeout);
+      return response.status === 200 ? response.json() : null;
```
Comment (yesil) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:129]: if the status is not 200, will `null` bill replaced with the `emptyEntitlements`?

**Review by yesil — COMMENTED** [review]
Unit tests for `getUserEntitlements.js` missing. Given that 100% test coverage is needed on Milo. I think there are multiple flows that need to be covered.

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js (hunk 3 above)
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js]: Nope, great catch - fixed.

**Comment by mokimo** [comment]
@yesil Yeah, I'm very aware the tests are missing (and I'm biting myself not just writing this test-driven to begin with) but we still got a few high priority topics to cover for the nav migration MVP and after those are done, we will start covering the whole new feds based global-navigation with tests

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 4]
```diff
@@ -63,27 +136,33 @@ const getSubscriptions = async ({ queryParams, locale }) => {
...
+const getUserEntitlements = async ({ params, locale, format } = {}) => {
+  if (!window.adobeIMS?.isSignedInUser()) return Promise.resolve(emptyEntitlements());
 
   const queryParams = getQueryParameters(params);
-  if (entitlements[queryParams]) return entitlements[queryParams];
+  if (entitlements[queryParams]) {
+    return format === 'raw'
+      ? entitlements[queryParams]
+      : entitlements[queryParams]
+        .then((res) => mapSubscriptionCodes(res));
```
Comment (overmyheadandbody) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:150]: Would we also need a `catch` statement here?

Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js:150]: I'll try to revisit this comment when I write some tests in the future so I can actually check this exact scenario.
Added it for now

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 5]
```diff
@@ -36,7 +37,75 @@ const getQueryParameters = (params) => {
...
+    if (offer.product_arrangement_code) {
+      arrangment_codes[offer.product_arrangement_code] = true;
+    }
+
+    if (Array.isArray(fulfilled_items)) {
+      fulfilled_items.forEach(({ code }) => {
+        if (code) {
+          fulfilled_codes[code] = true;
+          // Avoid duplicates
+          if (list.fulfilled_codes.indexOf(code) === -1) {
```
[the remark this answers is not in the document]
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js:88]: 1:1 ported that, I might slightly change it once I write tests for this so I'm sure nothing breaks by changing it right now

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 6]
```diff
@@ -36,7 +37,75 @@ const getQueryParameters = (params) => {
...
+    list,
+  } = emptyEntitlements();
+
+  offers.forEach(({ fulfilled_items, offer = {} }) => {
+    const cloud = offer.product_arrangement?.cloud;
+    clouds[CREATIVE_CLOUD] = clouds[CREATIVE_CLOUD] || cloud === 'CREATIVE';
+    clouds[DOCUMENT_CLOUD] = clouds[DOCUMENT_CLOUD] || cloud === 'DOCUMENT';
+    clouds[EXPERIENCE_CLOUD] = clouds[EXPERIENCE_CLOUD] || cloud === 'EXPERIENCE';
+
+    const family = offer?.product_arrangement?.family;
```
[the remark this answers is not in the document]
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js]: It will default to an empty object if not available, so we can safely remove it 👍

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 7]
```diff
@@ -36,7 +37,75 @@ const getQueryParameters = (params) => {
   return result;
 };
 
+const emptyEntitlements = () => ({
```
[the remark this answers is not in the document]
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js]: Nope that part was just ported from the aem version - I don't see any reason for it to stay a function - changed 👍

**Review by narcis-radu — APPROVED** [review]

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 8]
```diff
@@ -36,7 +37,75 @@ const getQueryParameters = (params) => {
...
+    offer_families,
+    list,
+  };
+};
+
 const getSubscriptions = async ({ queryParams, locale }) => {
+  const controller = new AbortController();
+  const { signal } = controller;
+  const timeout = setTimeout(() => controller.abort(), API_WAIT_TIMEOUT);
   const profile = await window.adobeIMS.getProfile();
```
Comment (honstar) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:109]: While this line is not part of this PR, I'm wondering whether we should guard against `window.adobeIMS` being null at this point?

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js [hunk 9]
```diff
@@ -36,7 +37,75 @@ const getQueryParameters = (params) => {
...
+    fulfilled_codes,
+    offer_families,
+    list,
+  };
+};
+
 const getSubscriptions = async ({ queryParams, locale }) => {
+  const controller = new AbortController();
+  const { signal } = controller;
+  const timeout = setTimeout(() => controller.abort(), API_WAIT_TIMEOUT);
```
Comment (honstar) [inline libs/blocks/global-navigation/utilities/getUserEntitlements.js:108]: Maybe it makes more sense to use `AbortSignal.timeout(API_WAIT_TIMEOUT)` instead of setting a custom timeout? See https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout

File: libs/blocks/global-navigation/utilities/getUserEntitlements.js (hunk 8 above)
Comment (mokimo) [reply libs/blocks/global-navigation/utilities/getUserEntitlements.js:109]: The `getUserEntitlements` method ensures IMS is available and the user is logged in 
```js
const getUserEntitlements = async ({ params, locale, format } = {}) => {
  if (!window.adobeIMS?.isSignedInUser()) return Promise.resolve(emptyEntitlements);
  ... 
  getSubscriptions()
```
`

**Review by overmyheadandbody — APPROVED** [review]
